### PR TITLE
Handle ANCM shutdown

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses #385. For manual testing, I removed the MS-ASPNETCORE-EVENT check and verified the shutdown mechanism.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs [hunk 1]
```diff
@@ -66,6 +74,17 @@ public IISMiddleware(RequestDelegate next, ILoggerFactory loggerFactory, IOption
                 return;
             }
 
+            // Handle shutdown from ANCM
+            if (HttpMethods.IsPost(httpContext.Request.Method) &&
+                string.Equals(ANCMShutdownRequestPath, httpContext.Request.Path.Value, StringComparison.OrdinalIgnoreCase) &&
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs]: The default for path comparison is `OrdinalIgnoreCase` https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Abstractions/PathString.cs#L332.

File: src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs [hunk 2]
```diff
@@ -66,6 +74,17 @@ public IISMiddleware(RequestDelegate next, ILoggerFactory loggerFactory, IOption
...
+                string.Equals(ANCMShutdownEventHeaderValue, httpContext.Request.Headers[MSAspNetCoreEvent], StringComparison.Ordinal))
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs]: ANCM is sending this so I think it's safe to use case-sensitive comparison. As a reference the token also uses case-sensitive compare.

File: src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs [hunk 3]
```diff
@@ -20,13 +19,17 @@ public class IISMiddleware
     {
         private const string MSAspNetCoreClientCert = "MS-ASPNETCORE-CLIENTCERT";
         private const string MSAspNetCoreToken = "MS-ASPNETCORE-TOKEN";
+        private const string MSAspNetCoreEvent = "MS-ASPNETCORE-EVENT";
+        private const string ANCMShutdownRequestPath = "/iisintegration";
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs]: ANCMRequestPath

File: src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs]: Use Request.Path.Equals

File: src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs (hunk 2 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs]: OrdinalIgnoreCase

File: src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs [hunk 4]
```diff
@@ -66,6 +74,17 @@ public IISMiddleware(RequestDelegate next, ILoggerFactory loggerFactory, IOption
...
             }
 
+            // Handle shutdown from ANCM
+            if (HttpMethods.IsPost(httpContext.Request.Method) &&
+                string.Equals(ANCMShutdownRequestPath, httpContext.Request.Path.Value, StringComparison.OrdinalIgnoreCase) &&
+                string.Equals(ANCMShutdownEventHeaderValue, httpContext.Request.Headers[MSAspNetCoreEvent], StringComparison.Ordinal))
+            {
+                // Execute shutdown task on background thread without waiting for completion
+                var shutdownTask = Task.Run(() => _applicationLifetime.StopApplication());
+                httpContext.Response.StatusCode = 202;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs]: Use the status codes constant

File: test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs [hunk 5]
```diff
@@ -73,6 +75,173 @@ public class IISMiddlewareTests
...
+                });
+            var server = new TestServer(builder);
+
+            var request = new HttpRequestMessage(HttpMethod.Post, "/iisintegration");
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-TOKEN", "TestToken");
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-EVENT", "shutdown");
+            var response = await server.CreateClient().SendAsync(request);
+
+            Assert.False(assertsExecuted);
+            await Task.Delay(TimeSpan.FromSeconds(1));
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs]: Added this to ensure test pass consistency, will remove task delay and use change tokens instead.

Comment (Tratcher) [inline test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs]: What's the delay for? Use a ManualResetEvent with a wait timeout if you're waiting for something specific to happen.

File: test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs [hunk 6]
```diff
@@ -73,6 +75,173 @@ public class IISMiddlewareTests
...
 
         [Fact]
+        public async Task MiddlewareShutsdownGivenANCMShutdown()
+        {
+            var assertsExecuted = false;
+            var applicationStoppingFired = false;
+            var builder = new WebHostBuilder()
+                .UseSetting("TOKEN", "TestToken")
+                .UseSetting("PORT", "12345")
+                .UseSetting("APPL_PATH", "/")
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs]: Add a test for PathBase. E.g. APPL_PATH = "/base"

File: src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs (hunk 2 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/IISMiddleware.cs]: You can use httpContext.Request.Headers[MSAspNetCoreEvent].Equals

**Comment by JunTaoLuo** [comment]
:up::date:

File: test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs [hunk 7]
```diff
@@ -72,6 +74,165 @@ public class IISMiddlewareTests
...
+                    });
+                });
+            var server = new TestServer(builder);
+
+            var request = new HttpRequestMessage(HttpMethod.Post, requestPath);
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-TOKEN", "TestToken");
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-EVENT", shutdownEvent);
+            var response = await server.CreateClient().SendAsync(request);
+
+            Assert.False(requestExecuted.WaitOne(0));
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs]: Check requestExecuted after checking applicationStoppingFired.

File: test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs [hunk 8]
```diff
@@ -72,6 +74,165 @@ public class IISMiddlewareTests
...
+                    });
+                });
+            var server = new TestServer(builder);
+
+            var request = new HttpRequestMessage(method, "/iisintegration");
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-TOKEN", "TestToken");
+            request.Headers.TryAddWithoutValidation("MS-ASPNETCORE-EVENT", "shutdown");
+            var response = await server.CreateClient().SendAsync(request);
+
+            Assert.True(requestExecuted.WaitOne(0));
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs]: 0 is weird for these, it assumes the request was executed synchronously. Use 5 seconds.

Comment (JunTaoLuo) [inline test/Microsoft.AspNetCore.Server.IISIntegration.Tests/IISMiddlewareTests.cs]: Gonna do 1 second. There are a lot of test and I don't want these tests to take forever. 1 second is probably enough.

**Pull request merged**